src/components/ChannelList.js

```diff
@@ -14,74 +19,224 @@ import { isPromise } from '../utils';
+
+    /** Function that overrides default behaviour when users gets added to a channel */
+    onAddedToChannel: PropTypes.func,
+    /** Function that overrides default behaviour when users gets removed from a channel */
+    onRemovedFromChannel: PropTypes.func,
+
+    /** Object containing query filters */
+    filters: PropTypes.object,
+    /** Object containing query options */
+    options: PropTypes.object,
+    /** Object containing sort parameters */
+    sort: PropTypes.object,
   };
 
   static defaultProps = {
     Preview: ChannelPreviewLastMessage,
     LoadingIndicator,
     List: ChannelListTeam,
+    Paginator: LoadMorePaginator,
+
+    filters: {},
+    options: {},
+    sort: {},
   };
 
   constructor(props) {
     super(props);
-    this.state = { error: false, loading: true, channels: [] };
+    this.state = {
+      // list of channels
+      channels: Immutable([]),
+      // loading channels
+      loadingChannels: true,
+      // error loading channels
+      refreshing: false,
+      offset: 0,
+      error: false,
+    };
 
     this.menuButton = React.createRef();
   }
 
+  static getDerivedStateFromError() {
+    return { error: true };
+  }
+
+  componentDidCatch(error, info) {
+    console.warn(error, info);
+  }
+
   async componentDidMount() {
+    await this.queryChannels();
+    this.listenToChanges();
+  }
+
+  queryChannels = async () => {
+    const { options, filters, sort } = this.props;
+    const { offset } = this.state;
+
+    this.setState({ refreshing: true });
+
+    const channelPromise = this.props.client.queryChannels(filters, sort, {
+      ...options,
+      offset,
+    });
+
     try {
-      let channelQueryResponse = this.props.channels;
+      let channelQueryResponse = channelPromise;
       if (isPromise(channelQueryResponse)) {
-        channelQueryResponse = await this.props.channels;
-        if (channelQueryResponse.length >= 1) {
+        channelQueryResponse = await channelPromise;
+        if (offset === 0 && channelQueryResponse.length >= 1) {
           this.props.setActiveChannel(channelQueryResponse[0]);
         }
       }
-      this.setState({ loading: false, channels: channelQueryResponse });
+      this.setState((prevState) => {
+        const channels = [...prevState.channels, ...channelQueryResponse];
+        return {
+          channels, // not unique somehow needs more checking
+          loadingChannels: false,
+          offset: channels.length,
+          hasNextPage:
+            channelQueryResponse.length >= options.limit ? true : false,
+          refreshing: false,
+        };
+      });
     } catch (e) {
-      console.log(e);
-      this.setState({ error: true });
+      console.warn(e);
+      this.setState({ error: true, refreshing: false });
     }
-  }
+  };
 
-  static getDerivedStateFromError() {
-    return { error: true };
+  listenToChanges() {
+    this.props.client.on(this.handleEvent);
```

Comment:
Better to switch it off in `componentWillUnmount`. It will cause memory leak if customer decides to keep `ChannelList` and `Channel/MessageList` component under separate `Chat` components and it involves lot of mounting and unmounting while switching between channellist and messagelist components, which is the case fpr react native.